### PR TITLE
[DebugInfo] Support translation of DebugEntryPoint instruction

### PR DESCRIPTION
**Pull request by vmaksimo**
This instruction is generated for DWARF `DISPFlagMainSubprogram` flag of function as well as for `spir_kernel` functions.

Spec:
https://github.com/KhronosGroup/SPIRV-Registry/blob/main/nonsemantic/NonSemantic.Shader.DebugInfo.100.asciidoc#DebugEntryPoint

### CODE REVIEW

**Comment by vmaksimo** [comment]
@MrSidims @asudarsa @maksimsab @jgstarIntel @LU-JOHN please take a look

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 1]
```diff
@@ -1196,6 +1216,10 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgFunction(const DISubprogram *Func) {
     DebugFunc = transDbgTemplateParams(TPA, DebugFunc);
   }
 
+  if (isNonSemanticDebugInfo() &&
+      (Func->isMainSubprogram() || IsEntryPointKernel))
+    transDbgEntryPoint(Func, DebugFunc);
```
Comment (MrSidims) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp:1221]: I just realized, that for this case and for the case below - transDbgFuncDefinition, static analyzer tools will complain, that result of the functions is unused. Can we make these two functions returning void? (can be done it this PR)

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 2]
```diff
@@ -1219,6 +1243,28 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgFuncDefinition(SPIRVValue *FuncDef,
                         Ops, BB, BB->getInst(0));
 }
 
+SPIRVEntry *LLVMToSPIRVDbgTran::transDbgEntryPoint(const DISubprogram *Func,
+                                                   SPIRVEntry *DbgFunc) {
+  if (!isNonSemanticDebugInfo())
```
Comment (MrSidims) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp:1248]: We've already checked isNonSemanticDebugInfo before

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 3]
```diff
@@ -1219,6 +1243,28 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgFuncDefinition(SPIRVValue *FuncDef,
...
+  SPIRVWordVec Ops(SPIRVDebug::Operand::EntryPoint::OperandCount);
+
+  DICompileUnit *CU = Func->getUnit();
+  StringRef Producer = CU->getProducer();
+  StringRef Flags = CU->getFlags();
+
+  auto It = MDMap.find(CU);
+  SPIRVEntry *CUVal = It != MDMap.end() ? It->second : getDebugInfoNone();
+
+  using namespace SPIRVDebug::Operand::EntryPoint;
```
Comment (MrSidims) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp:1260]: nit:
can be moved before
`SPIRVWordVec Ops(SPIRVDebug::Operand::EntryPoint::OperandCount);`

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp [hunk 4]
```diff
@@ -1219,6 +1243,28 @@ SPIRVEntry *LLVMToSPIRVDbgTran::transDbgFuncDefinition(SPIRVValue *FuncDef,
...
+    return nullptr;
+
+  SPIRVWordVec Ops(SPIRVDebug::Operand::EntryPoint::OperandCount);
+
+  DICompileUnit *CU = Func->getUnit();
+  StringRef Producer = CU->getProducer();
+  StringRef Flags = CU->getFlags();
+
+  auto It = MDMap.find(CU);
+  SPIRVEntry *CUVal = It != MDMap.end() ? It->second : getDebugInfoNone();
```
Comment (MrSidims) [inline lib/SPIRV/LLVMToSPIRVDbgTran.cpp:1258]: We can probably get use of brand-new SPIRVCUMap

File: lib/SPIRV/SPIRVToLLVMDbgTran.cpp [hunk 5]
```diff
@@ -909,6 +912,18 @@ DINode *SPIRVToLLVMDbgTran::transFunctionDecl(const SPIRVExtInst *DebugInst) {
   return DIS;
 }
 
+MDNode *SPIRVToLLVMDbgTran::transEntryPoint(const SPIRVExtInst *DebugInst) {
+  if (!isNonSemanticDebugInfo(DebugInst->getExtSetKind()))
```
Comment (MrSidims) [inline lib/SPIRV/SPIRVToLLVMDbgTran.cpp:916]: Is it a possible scenario? If we believe it's not - better to convert to assertion.

File: lib/SPIRV/libSPIRV/SPIRV.debug.h [hunk 6]
```diff
@@ -55,6 +55,7 @@ enum Instruction {
   ModuleINTEL                   = 36,
   InstCount                     = 37,
   FunctionDefinition            = 101,
+  EntryPoint                    = 107,
```
[the remark this answers is not in the document]
Comment (vmaksimo) [reply lib/SPIRV/libSPIRV/SPIRV.debug.h:58]: Should I have done this for FunctionDefinition also?

[the remark this answers is not in the document]
Comment (MrSidims) [reply lib/SPIRV/libSPIRV/SPIRV.debug.h:58]: No, FunctionDefinition has no parent scope, aka it's linked with CU implicitly thorough the appropriate DebugFunction

File: lib/SPIRV/LLVMToSPIRVDbgTran.cpp (hunk 1 above)
Comment (vmaksimo) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp:1221]: It's possible to make these two functions return void. However they call `BM->addDebugInfo` inside, which does return the value - so the problem for static analysis will still exist. Not sure how to resolve this properly

Comment (MrSidims) [reply lib/SPIRV/LLVMToSPIRVDbgTran.cpp:1221]: okay, lets do not change result type to void. Instead we can add
 [[maybe_unused]] auto *Inst = transDbgEntryPoint

**Review by MrSidims — APPROVED** [review]
LGTM, thanks!
Please fix clang-format

**Pull request merged**